### DISCUSSION TITLE
VRPTW: slack time should vary based on load

### DISCUSSION THREAD

**Discussion by odashevsky** [google/or-tools#2857] [Routing (and legacy CP) questions] [answered]
Hi,

Solving a VRPTW problem and am facing a requirement where a vehicle can idle ("slack") only Y when empty, but if it has at least one load onboard it should not idle waiting for a second pickup more than Y. We are specifying slack time using this function:

`public bool AddDimension(int evaluator_index, long slack_max, long capacity, bool fix_start_cumul_to_zero, string name);`

But how would I go about "instructing" the solver "slack_max" should be shorted if the vehicle is not empty?

Thank you.

**Comment by Mizux** [comment, marked as answer]
```python
solver = routing.solver()
solver.Add(time_dim.SlackVar(index) <= load_dim.CumulVar(index) * 1000)
```
 ?

**Reply by odashevsky** [reply]
Hi Mizux, thanks for the suggestion. Wouldn't the suggested constraint make the slack variable 0 (no slack) is vehicle is empty and will make the slack larger and larger as the vehicle gets loaded more? My requirement is the opposite: slack is large if empty and gets down to, for example, no more than 15 units if not-empty.
Thanks

**Reply by Mizux** [reply]
erf good point, my bad, miss read..
```python
solver = routing.solver()
tmp = load_dim.CumulVar(index) > 0 # create a BooleanVar true aka 1 if non empty
solver.Add(time_dim.SlackVar(index) * tmp <= 15 * tmp)
```

if `empty` then tmp == 0 so we have constraint `0 <= 0` i.e. disabled.
if `non empty` then tmp == 1 so we have constraint `SlackVar(index) <= 15*1` i.e. no more than 15min (supposing the time dimension use minute as unit).

**Reply by Mizux** [reply]
if you want a limit in both case you could use.
```py
solver.Add(time_dim.SlackVar(index) <= tmp * 15 + (1 - tmp) * 45)
```
aka max 15min if loaded otherwise 45min...

**Reply by odashevsky** [reply]
Thanks! I will give it a try.

**Reply by odashevsky** [reply]
If only you could give some help with [(https://github.com/google/or-tools/discussions/2850)]  
that'd be awesome :):)

**Comment by jmarca** [comment]
Jumping in on this discussion, I’ve often wondered about the performance of the solver in theses sorts of cases. The slack here is piecewise linear, so does the solver recognize that and adjust accordingly? Or does it make no difference?
